### PR TITLE
[cxx-interop] Allow compiling with libc++ on Linux

### PR DESCRIPTION
**Pull request by egorzhdan**
This makes sure that Swift respects `-Xcc -stdlib=libc++` flags.

Clang already has existing logic to discover the system-wide libc++ installation on Linux. We rely on that logic here.

Importing a Swift module that was built with a different C++ stdlib is not supported and emits an error.

The Cxx module can be imported when compiling with any C++ stdlib. The synthesized conformances, e.g. to CxxRandomAccessCollection also work. However, CxxStdlib currently cannot be imported when compiling with libc++, since on Linux it refers to symbols from libstdc++ which have different mangled names in libc++.

rdar://118357548 / https://github.com/swiftlang/swift/issues/69825

### CODE REVIEW

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

File: include/swift/AST/DiagnosticsSema.def [hunk 1]
```diff
@@ -942,6 +942,13 @@ ERROR(need_cxx_interop_to_import_module,none,
 NOTE(enable_cxx_interop_docs,none,
      "visit https://www.swift.org/documentation/cxx-interop/project-build-setup to learn how to enable C++ interoperability", ())
 
+ERROR(cxx_stdlib_kind_mismatch,none,
+      "module %0 was built with "
+      "%select{platform default C++ stdlib|libc++}1, "
+      "but current compilation uses "
+      "%select{platform default C++ stdlib|libc++}2",
```
Comment (compnerd) [inline include/swift/AST/DiagnosticsSema.def:949]: I'm not sure that this is great - the platform default c++ runtime could be libc++. I'd rather that we use a string parameter and let the frontend specify the runtime name.

File: lib/Frontend/CompilerInvocation.cpp [hunk 2]
```diff
@@ -627,6 +627,16 @@ void LangOptions::setCxxInteropFromArgs(ArgList &Args,
...
+
+  for (const Arg *A : Args.filtered(options::OPT_Xcc)) {
+    StringRef clangArg = A->getValue();
+    if (clangArg.consume_front("-stdlib")) {
+      clangArg.consume_front("=");
+      if (clangArg.equals("libc++"))
+        CXXStdlib = CXXStdlibKind::OverrideLibcxx;
+      // If the stdlib argument is invalid, Clang will emit an error.
+    }
+  }
```
Comment (compnerd) [inline lib/Frontend/CompilerInvocation.cpp:639]: I wonder if we should just let clang parse these and then pull it from the clang importer.

File: include/swift/Basic/CXXStdlibKind.h [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+#ifndef SWIFT_BASIC_CXX_STDLIB_KIND_H
+#define SWIFT_BASIC_CXX_STDLIB_KIND_H
+
+namespace swift {
+
+enum class CXXStdlibKind : uint8_t {
+  /// Use the default C++ stdlib for a particular platform: libc++ for Darwin,
+  /// libstdc++ for most Linux distros.
+  PlatformDefault = 0,
```
Comment (compnerd) [inline include/swift/Basic/CXXStdlibKind.h:21]: I think that we shouldn't just use a blanket "PlatformDefault" .. what if I want to use libstdc++ on a libc++ defaulting platform? We really should enumerate the C++ runtimes IMO:

1. msvcprt
2. libstdc++
3. libc++
4. stl_port
5. stdc++
6. dinkumware
7. eastl

are ones off the top of my head that I can think of as being quite common.

Comment (etcwilde) [inline include/swift/Basic/CXXStdlibKind.h:21]: I'm not a huge fan of embedding "platform default" vs "libcxx" in the compiler given that our notion of "default" isn't necessarily right. Android is technically a Linux, where the NDK provides a libcxx to link against. But, there isn't a standard system-wide C++ runtime either since Google doesn't guarantee C++ ABI stability, it's whatever was shipped in the NDK. That also means that I could totally use libstdc++ with my program if I wanted to and neither is the default.

should this enum instead list out the supported C++ runtimes (which are enumerable since we also have to provide the modulemap bits anyway). Then it's less prescriptive, and we could even consider exposing a mechanism for the vendor and the SDK to say what their "default" C++ runtime is.

File: lib/Frontend/CompilerInvocation.cpp (hunk 2 above)
Comment (egorzhdan) [reply lib/Frontend/CompilerInvocation.cpp:639]: Yeah, I initially thought of doing that in a similar way to how `getLibStdCxxFileMapping` extracts the libstdc++ installation path. This solution seemed to require a decent amount of plumbing through ClangImporter to get the stdlib kind to LangOptions, but let me see if I can do it in a simpler way.

File: include/swift/Basic/CXXStdlibKind.h (hunk 3 above)
Comment (egorzhdan) [reply include/swift/Basic/CXXStdlibKind.h:21]: I'd like to keep this patch focused on a single use case, libc++ on Linuxes that use libstdc++ by default. I'm trying to keep the implementation fairly generic, to make support of other stdlibs a potential future improvement. But I'm not familiar with the rest of the libraries you've listed and I don't really have an easy way to test them.

> what if I want to use libstdc++ on a libc++ defaulting platform?

In that case we would need to add `OverrideLibstdcxx` to this enum.

Comment (egorzhdan) [reply include/swift/Basic/CXXStdlibKind.h:21]: I think a reasonable definition of the "default" here would be: what library Clang picks when you run `clang++ x.cpp`. Having a different default than Clang would be quite disastrous since people would try to build a C++ library with Clang (without passing `-stdlib=xyz`) and then use it from Swift (again, without passing `-Xcc -stdlib=xyz`) and not realize that they're using two different C++ stdlibs.

On Android, Clang uses libc++ by default, and `PlatformDefault` would mean libc++ when targeting Android.

The actual meaning of `PlatformDefault` here is that the library on client's side is the same as the library that Swift was built against, meaning the compiler is free to use prebuilt binaries such as the CxxStdlib overlays.

File: lib/Frontend/CompilerInvocation.cpp (hunk 2 above)
Comment (egorzhdan) [reply lib/Frontend/CompilerInvocation.cpp:639]: The main issue with extracting the stdlib kind from Clang seems to be Windows support: Clang doesn't seem to be aware that the Microsoft standard library is its own thing. It assumes that the default on Windows is libstdc++. It also cannot discover libc++ on Windows: 

https://github.com/llvm/llvm-project/blob/68df06a0b2998765cb0a41353fcf0919bbf57ddb/clang/lib/Driver/ToolChains/MSVC.cpp#L776-L779

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

File: include/swift/Basic/CXXStdlibKind.h (hunk 3 above)
Comment (compnerd) [reply include/swift/Basic/CXXStdlibKind.h:21]: Having them be enumerated and not used should be fine though. I don't think that you need to implement support for all of them. I'm just trying to show that "platform default" doesn't convey enough information.

File: lib/Frontend/CompilerInvocation.cpp (hunk 2 above)
Comment (compnerd) [reply lib/Frontend/CompilerInvocation.cpp:639]: We can map that as a workaround, but we should really try to map the runtime rather than claim "default". What if there is no default (e.g. android)?

File: include/swift/Basic/CXXStdlibKind.h (hunk 3 above)
Comment (compnerd) [reply include/swift/Basic/CXXStdlibKind.h:21]: Clang now uses libc++ by default, but if you change the platform version, the default should be gnustl (aka libstdc++). They also recommended stlport for a while. I think that "platform default" just doesn't really give you much information (especially as a user).

Comment (egorzhdan) [reply include/swift/Basic/CXXStdlibKind.h:21]: From what I can tell, Clang seems to use libc++ as the default on Android regardless of the version: https://github.com/llvm/llvm-project/blob/bc747c3e1377e3e262d901cefcae2f563c895bb5/clang/lib/Driver/ToolChains/Linux.cpp#L367-L368

File: lib/Frontend/CompilerInvocation.cpp (hunk 2 above)
Comment (egorzhdan) [reply lib/Frontend/CompilerInvocation.cpp:639]: (I would consider libc++ to be the default on Android – see the thread below)

Comment (egorzhdan) [reply lib/Frontend/CompilerInvocation.cpp]: Done, I removed the manual parsing logic and switched to extracting the C++ stdlib kind from Clang Driver.

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

File: include/swift/Basic/CXXStdlibKind.h (hunk 3 above)
Comment (compnerd) [reply include/swift/Basic/CXXStdlibKind.h:21]: How about renaming `PlatformDefault` to `ClangDefault`?

File: lib/Serialization/SerializedModuleLoader.cpp [hunk 4]
```diff
@@ -1055,6 +1057,17 @@ LoadedFile *SerializedModuleLoaderBase::loadAST(
                        M.getName());
     Ctx.Diags.diagnose(loc, diag::enable_cxx_interop_docs);
   }
+  // Modules built with libc++ cannot be imported into modules that are built
+  // with libstdc++, and vice versa. Make an exception for Cxx.swiftmodule since
+  // it doesn't refer to any C++ stdlib symbols, and for CxxStdlib.swiftmodule
+  // since we skipped loading the overlay for the module.
+  if (M.hasCxxInteroperability() && Ctx.LangOpts.EnableCXXInterop &&
+      M.getCXXStdlibKind() != Ctx.LangOpts.CXXStdlib &&
+      M.getName() != Ctx.Id_Cxx && M.getName() != Ctx.Id_CxxStdlib) {
```
Comment (compnerd) [inline lib/Serialization/SerializedModuleLoader.cpp:1066]: What about unconditionally encoding the C++ runtime into the module if C++ Interop is enabled?

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Review by al45tair — COMMENTED** [review]
I think we should be using more than one bit in the `swiftmodule` file, and that we should identify _which_ C++ library the module was built with (rather than "default" versus "overridden", which is what we have now).

File: include/swift/AST/DiagnosticsSema.def (hunk 1 above)
Comment (egorzhdan) [reply include/swift/AST/DiagnosticsSema.def]: I changed this to explicitly list the supported C++ stdlibs by name.

File: include/swift/Basic/CXXStdlibKind.h (hunk 3 above)
Comment (egorzhdan) [reply include/swift/Basic/CXXStdlibKind.h]: I changed this to list the supported C++ stdlibs explicitly: libc++, libstdc++, msvcprt. I also removed the platform-default option to make this logic easier to understand.

I agree that we shouldn't list the stdlibs that we don't actually support.

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

File: lib/Serialization/SerializedModuleLoader.cpp (hunk 4 above)
Comment (egorzhdan) [reply lib/Serialization/SerializedModuleLoader.cpp:1066]: We are generally doing that, for all modules except Cxx and CxxStdlib. The Cxx module doesn't refer to any C++ stdlib APIs despite being built with C++ interop enabled. The CxxStdlib module has an overlay that is handled separately: ClangImporter skips loading the overlay for CxxStdlib whenever compiling with a non-default C++ stdlib.

**Comment by hyp** [comment]
I tested this on windows, with some additional changes for custom libc++ and split out MSVC c++ support using sealing like before, and this kind of approach works great. This looks good to me.

**Comment by egorzhdan** [comment]
Thanks @hyp for trying it out!

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please build toolchain Ubuntu 24.04

**Comment by egorzhdan** [comment]
@swift-ci please build toolchain Ubuntu 23.10

**Comment by egorzhdan** [comment]
@swift-ci please build toolchain Ubuntu 22.04

File: include/swift/AST/DiagnosticsSema.def [hunk 5]
```diff
@@ -942,6 +942,13 @@ ERROR(need_cxx_interop_to_import_module,none,
 NOTE(enable_cxx_interop_docs,none,
      "visit https://www.swift.org/documentation/cxx-interop/project-build-setup to learn how to enable C++ interoperability", ())
 
+ERROR(cxx_stdlib_kind_mismatch,none,
+      "module %0 was built with "
+      "%select{unknown C++ stdlib|libc++|libstdc++|msvcprt}1, "
+      "but current compilation uses "
+      "%select{unknown C++ stdlib|libc++|libstdc++|msvcprt}2",
+      (Identifier, CXXStdlibKind, CXXStdlibKind))
```
Comment (compnerd) [inline include/swift/AST/DiagnosticsSema.def:950]: Might be nice to use a to_string to get the runtime name should we support additional ones.

File: include/swift/AST/Decl.h [hunk 6]
```diff
@@ -780,6 +780,10 @@ class alignas(1 << DeclAlignInBits) Decl : public ASTAllocated<Decl> {
     /// Whether this module has been built with C++ interoperability enabled.
     HasCxxInteroperability : 1,
 
+    /// Whether this module uses the platform default C++ stdlib, or an
+    /// overridden C++ stdlib.
+    CXXStdlibKind : 2,
```
Comment (compnerd) [inline include/swift/AST/Decl.h:785]: This can lead to truncation of we add all enumerator. Should we just hold the space?

File: lib/IRGen/IRGenModule.cpp [hunk 7]
```diff
@@ -1646,6 +1646,7 @@ void IRGenModule::addLinkLibraries() {
                                               getSwiftModule()->getName().str(),
                                               hasStaticCxx,
                                               hasStaticCxxStdlib,
+                                              Context.LangOpts.CXXStdlib,
```
Comment (compnerd) [inline lib/IRGen/IRGenModule.cpp:1649]: Unnecessary change?

**Review by compnerd — APPROVED** [review]
Thank you for all the iteration. This really seems significantly better than both the first revision and even some of the original code.

File: include/swift/AST/Decl.h (hunk 6 above)
Comment (egorzhdan) [reply include/swift/AST/Decl.h:785]: Do you mean we should reserve more bits for this field?

File: lib/IRGen/IRGenModule.cpp (hunk 7 above)
Comment (egorzhdan) [reply lib/IRGen/IRGenModule.cpp:1649]: We do use this value in `registerCxxInteropLibraries` to determine which C++ stdlib be should link against.

File: include/swift/AST/DiagnosticsSema.def (hunk 5 above)
Comment (egorzhdan) [reply include/swift/AST/DiagnosticsSema.def]: Good point, I switched this to use `to_string`.

File: include/swift/AST/Decl.h (hunk 6 above)
Comment (al45tair) [reply include/swift/AST/Decl.h:785]: On the one hand, this is versioned, but on the other hand reserving (say) eight bits up front means we would likely never run into a problem that required us to update it. We would need to be a little careful about decoding unknown values for it, potentially, but I can't see a real downside.

**Review by al45tair — APPROVED** [review]

**Comment by egorzhdan** [comment]
@swift-ci please smoke test

**Comment by egorzhdan** [comment]
@swift-ci please build toolchain UBI9

File: lib/IRGen/IRGenModule.cpp (hunk 7 above)
Comment (compnerd) [reply lib/IRGen/IRGenModule.cpp:1649]: Hmm, weird, it showed up a whitespace only change 😕
